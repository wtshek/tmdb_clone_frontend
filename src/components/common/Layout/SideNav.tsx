import Typography, {
  FontSizeEnum,
  TextColorEnum,
} from '@/components/common/Typography/Typography';

export enum NavItemType {
  MAIN = 'main',
  SUB = 'sub',
}

export type NavItem = {
  type: NavItemType;
  key: string;
  label: string;
  href: string;
};

type SideNavProps = {
  isOpen: boolean;
  items: NavItem[];
};

const SideNav = ({ isOpen, items }: SideNavProps) => {
  return (
    <div className="relative">
      {/* Menu that slides in from the left */}
      <div
        className={`fixed top-16 left-0 h-screen w-64 bg-blue-blur backdrop-blur-[20px]  z-50 transition-transform duration-300 transform ${
          isOpen ? 'translate-x-0' : '-translate-x-full'
        }`}
      >
        {/* Menu contents */}
        <nav className="py-8 px-4">
          <ul>
            {items.map((item) => {
              const { key, label, type } = item;
              const isMain = type === NavItemType.MAIN;
              return (
                <li key={key} className="mt-2.5">
                  <Typography
                    textColor={
                      isMain ? TextColorEnum.WHITE : TextColorEnum.GRAY
                    }
                    fontSize={isMain ? FontSizeEnum.XL : FontSizeEnum.BASE}
                  >
                    {label}
                  </Typography>
                </li>
              );
            })}
          </ul>
        </nav>
      </div>
    </div>
  );
};

export default SideNav;
