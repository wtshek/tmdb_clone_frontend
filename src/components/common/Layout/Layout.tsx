import { GiHamburgerMenu } from 'react-icons/gi';
import { BsBellFill } from 'react-icons/bs';
import { MdOutlineSettingsPower } from 'react-icons/md';
import { BiSearch } from 'react-icons/bi';
import Image from 'next/image';
import { IconContext } from 'react-icons/lib';
import {
  FontSizeEnum,
  FontWeightEnum,
  TextColorEnum,
  Typography,
} from '@/components/common/Typography';
import SideNav, { NavItem, NavItemType } from './SideNav';
import { useState } from 'react';

type LayoutProps = {
  children: JSX.Element;
};

const LOGO_WIDTH = 55;
const LOGO_HEIGHT = 44;
const ICON_SIZE = '22px';
const ICON_COLOR = 'white';
const SEARCH_ICON_COLOR = '#10B5E5';

const footer = {
  theBasics: {
    header: 'the basics',
    items: [
      {
        label: 'About TMDB',
        href: '/about-tmdb',
      },
      {
        label: 'Contact Us',
        href: '/contact-us',
      },
      {
        label: 'Support Forums',
        href: '/support-forums',
      },
      {
        label: 'API',
        href: '/api',
      },
      {
        label: 'System Status',
        href: '/system-status',
      },
    ],
  },
  getInvolved: {
    header: 'get involved',
    items: [
      {
        label: 'Contribution Bible',
        href: '/contribution-bible',
      },
      {
        label: 'Add New Movie',
        href: '/add-new-movie',
      },
      {
        label: 'Add New Tv Show',
        href: '/add-new-tv-show',
      },
    ],
  },
  community: {
    header: 'Community',
    items: [
      {
        label: 'Guidelines',
        href: '/guidelines',
      },
      {
        label: 'Discussion',
        href: '/discussion',
      },
      {
        label: 'Leaderboard',
        href: '/leaderboard',
      },
      {
        label: 'Follow us on Twitter',
        href: 'https://twitter.com/TMDB',
      },
    ],
  },
  legal: {
    header: 'legal',
    items: [
      {
        label: 'Terms of Use',
        href: '/terms-of-use',
      },
      {
        label: 'API Terms of Use',
        href: '/api-terms-of-use',
      },
      {
        label: 'Privacy Policy',
        href: '/privacy-policy',
      },
    ],
  },
};

const sideNav: NavItem[] = [
  {
    type: NavItemType.MAIN,
    key: 'movies',
    label: 'Movies',
    href: '/movies',
  },
  {
    type: NavItemType.MAIN,
    key: 'tvshows',
    label: 'TV Shows',
    href: '/tvshows',
  },
  {
    type: NavItemType.MAIN,
    key: 'people',
    label: 'People',
    href: '/people',
  },
  {
    type: NavItemType.SUB,
    key: 'contributionBible',
    label: 'Contribution Bible',
    href: '/contribution-bible',
  },
  {
    type: NavItemType.SUB,
    key: 'apps',
    label: 'Apps',
    href: '/apps',
  },
  {
    type: NavItemType.SUB,
    key: 'discussion',
    label: 'Discussion',
    href: '/discussion',
  },
  {
    type: NavItemType.SUB,
    key: 'leaderboard',
    label: 'Leaderboard',
    href: '/leaderboard',
  },
  {
    type: NavItemType.SUB,
    key: 'api',
    label: 'API',
    href: '/api',
  },
  {
    type: NavItemType.SUB,
    key: 'support',
    label: 'Support',
    href: '/support',
  },
  {
    type: NavItemType.SUB,
    key: 'about',
    label: 'About',
    href: '/about',
  },
  {
    type: NavItemType.SUB,
    key: 'logout',
    label: 'Logout',
    href: '/logout',
  },
];

export const Layout = ({ children }: LayoutProps) => {
  const [isSideNavOpen, setIsSideNavOpen] = useState(false);

  const toggleSideNav = (): void => {
    setIsSideNavOpen((prev) => !prev);
  };

  return (
    <>
      <IconContext.Provider value={{ color: ICON_COLOR, size: ICON_SIZE }}>
        <nav className="fix flex justify-between bg-blue w-screen items-center px-5 h-16">
          <button className="w-1/3" onClick={toggleSideNav}>
            <GiHamburgerMenu />
          </button>
          <div className="flex justify-center w-1/3">
            <Image
              src="/tmdb_mobile.svg"
              alt="tmdb mobile logo"
              width={LOGO_WIDTH}
              height={LOGO_HEIGHT}
            />
          </div>
          <div className="flex justify-end w-1/3">
            <BsBellFill />
            <MdOutlineSettingsPower className="ml-4" />
            <IconContext.Provider
              value={{ color: SEARCH_ICON_COLOR, size: ICON_SIZE }}
            >
              <BiSearch className="ml-4" />
            </IconContext.Provider>
          </div>
        </nav>
        {children}
        <footer className="bg-blue py-10 px-5 md:flex md:justify-between md:px-16">
          {Object.entries(footer).map(([key, values]) => {
            const { header, items } = values;
            return (
              <div key={key} className="mb-8">
                <Typography
                  fontSize={FontSizeEnum.LG}
                  textColor={TextColorEnum.WHITE}
                  fontWeight={FontWeightEnum.BOLD}
                >
                  {header.toUpperCase()}
                </Typography>
                {items.map(({ label, href }) => (
                  <Typography textColor={TextColorEnum.WHITE} key={label}>
                    {label}
                  </Typography>
                ))}
              </div>
            );
          })}
        </footer>
        <SideNav isOpen={isSideNavOpen} items={sideNav} />
      </IconContext.Provider>
    </>
  );
};

export default Layout;
