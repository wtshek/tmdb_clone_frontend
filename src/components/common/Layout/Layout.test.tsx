import { render } from '@testing-library/react';
import Layout from './Layout';

describe('Layout component', () => {
  it('should render children and footer', () => {
    const { container } = render(
      <Layout>
        <div>Test</div>
      </Layout>
    );
    expect(container).toMatchSnapshot();
  });
});
