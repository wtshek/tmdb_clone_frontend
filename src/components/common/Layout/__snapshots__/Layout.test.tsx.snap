// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Layout component should render children and footer 1`] = `
<div>
  <nav
    class="fix flex justify-between bg-blue w-screen items-center px-5 h-16"
  >
    <button
      class="w-1/3"
    >
      <svg
        fill="currentColor"
        height="22px"
        stroke="currentColor"
        stroke-width="0"
        style="color: white;"
        viewBox="0 0 512 512"
        width="22px"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M32 96v64h448V96H32zm0 128v64h448v-64H32zm0 128v64h448v-64H32z"
        />
      </svg>
    </button>
    <div
      class="flex justify-center w-1/3"
    >
      <img
        alt="tmdb mobile logo"
        data-nimg="1"
        decoding="async"
        height="44"
        loading="lazy"
        src="/tmdb_mobile.svg"
        style="color: transparent;"
        width="55"
      />
    </div>
    <div
      class="flex justify-between w-1/3"
    >
      <svg
        fill="currentColor"
        height="22px"
        stroke="currentColor"
        stroke-width="0"
        style="color: white;"
        viewBox="0 0 16 16"
        width="22px"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M8 16a2 2 0 0 0 2-2H6a2 2 0 0 0 2 2zm.995-14.901a1 1 0 1 0-1.99 0A5.002 5.002 0 0 0 3 6c0 1.098-.5 6-2 7h14c-1.5-1-2-5.902-2-7 0-2.42-1.72-4.44-4.005-4.901z"
        />
      </svg>
      <svg
        fill="currentColor"
        height="22px"
        stroke="currentColor"
        stroke-width="0"
        style="color: white;"
        viewBox="0 0 24 24"
        width="22px"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 0h24v24H0V0z"
          fill="none"
        />
        <path
          d="M7 24h2v-2H7v2zm4 0h2v-2h-2v2zm2-22h-2v10h2V2zm3.56 2.44l-1.45 1.45A5.969 5.969 0 0118 11c0 3.31-2.69 6-6 6s-6-2.69-6-6c0-2.17 1.16-4.06 2.88-5.12L7.44 4.44A7.961 7.961 0 004 11c0 4.42 3.58 8 8 8s8-3.58 8-8c0-2.72-1.36-5.12-3.44-6.56zM15 24h2v-2h-2v2z"
        />
      </svg>
      <svg
        fill="currentColor"
        height="22px"
        stroke="currentColor"
        stroke-width="0"
        style="color: rgb(16, 181, 229);"
        viewBox="0 0 24 24"
        width="22px"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M10 18a7.952 7.952 0 0 0 4.897-1.688l4.396 4.396 1.414-1.414-4.396-4.396A7.952 7.952 0 0 0 18 10c0-4.411-3.589-8-8-8s-8 3.589-8 8 3.589 8 8 8zm0-14c3.309 0 6 2.691 6 6s-2.691 6-6 6-6-2.691-6-6 2.691-6 6-6z"
        />
      </svg>
    </div>
  </nav>
  <div>
    Test
  </div>
  <footer
    class="bg-blue py-10 px-5"
  >
    <div
      class="mb-8"
    >
      <div
        class="text-white text-lg font-bold"
      >
        THE BASICS
      </div>
      <div
        class="text-white text-base font-normal"
      >
        About TMDB
      </div>
      <div
        class="text-white text-base font-normal"
      >
        Contact Us
      </div>
      <div
        class="text-white text-base font-normal"
      >
        Support Forums
      </div>
      <div
        class="text-white text-base font-normal"
      >
        API
      </div>
      <div
        class="text-white text-base font-normal"
      >
        System Status
      </div>
    </div>
    <div
      class="mb-8"
    >
      <div
        class="text-white text-lg font-bold"
      >
        GET INVOLVED
      </div>
      <div
        class="text-white text-base font-normal"
      >
        Contribution Bible
      </div>
      <div
        class="text-white text-base font-normal"
      >
        Add New Movie
      </div>
      <div
        class="text-white text-base font-normal"
      >
        Add New Tv Show
      </div>
    </div>
    <div
      class="mb-8"
    >
      <div
        class="text-white text-lg font-bold"
      >
        COMMUNITY
      </div>
      <div
        class="text-white text-base font-normal"
      >
        Guidelines
      </div>
      <div
        class="text-white text-base font-normal"
      >
        Discussion
      </div>
      <div
        class="text-white text-base font-normal"
      >
        Leaderboard
      </div>
      <div
        class="text-white text-base font-normal"
      >
        Follow us on Twitter
      </div>
    </div>
    <div
      class="mb-8"
    >
      <div
        class="text-white text-lg font-bold"
      >
        LEGAL
      </div>
      <div
        class="text-white text-base font-normal"
      >
        Terms of Use
      </div>
      <div
        class="text-white text-base font-normal"
      >
        API Terms of Use
      </div>
      <div
        class="text-white text-base font-normal"
      >
        Privacy Policy
      </div>
    </div>
  </footer>
  <div
    class="relative"
  >
    <div
      class="fixed top-16 left-0 h-screen w-64 bg-blue-blur backdrop-blur-[20px]  z-50 transition-transform duration-300 transform -translate-x-full"
    >
      <nav
        class="py-8 px-4"
      >
        <ul>
          <li
            class="mt-2.5"
          >
            <div
              class="text-white text-xl font-normal"
            >
              Movies
            </div>
          </li>
          <li
            class="mt-2.5"
          >
            <div
              class="text-white text-xl font-normal"
            >
              TV Shows
            </div>
          </li>
          <li
            class="mt-2.5"
          >
            <div
              class="text-white text-xl font-normal"
            >
              People
            </div>
          </li>
          <li
            class="mt-2.5"
          >
            <div
              class="text-gray text-base font-normal"
            >
              Contribution Bible
            </div>
          </li>
          <li
            class="mt-2.5"
          >
            <div
              class="text-gray text-base font-normal"
            >
              Apps
            </div>
          </li>
          <li
            class="mt-2.5"
          >
            <div
              class="text-gray text-base font-normal"
            >
              Discussion
            </div>
          </li>
          <li
            class="mt-2.5"
          >
            <div
              class="text-gray text-base font-normal"
            >
              Leaderboard
            </div>
          </li>
          <li
            class="mt-2.5"
          >
            <div
              class="text-gray text-base font-normal"
            >
              API
            </div>
          </li>
          <li
            class="mt-2.5"
          >
            <div
              class="text-gray text-base font-normal"
            >
              Support
            </div>
          </li>
          <li
            class="mt-2.5"
          >
            <div
              class="text-gray text-base font-normal"
            >
              About
            </div>
          </li>
          <li
            class="mt-2.5"
          >
            <div
              class="text-gray text-base font-normal"
            >
              Logout
            </div>
          </li>
        </ul>
      </nav>
    </div>
  </div>
</div>
`;
